examples/rlhf/train_rlhf.py

```diff
@@ -0,0 +1,385 @@
+            )
+            logger.debug(string_to_write)
+
+        return test_reward
+
+    return estimate_loss
+
+
+def main():
+    config = load_config("config/train_rlhf.yaml")
+
+    query_logger = get_file_logger("query_logger", "query_logger.log")
+    test_reward_logger = get_file_logger("test_reward_logger", "test_rewards.log")
+
+    tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
+    tokenizer.pad_token = tokenizer.eos_token
+
+    reward_model = init_reward_model(config)
+    actor, critic, critic_head, model = init_actor_critic(config)
+    critic.eval()
+
+    ref_model = deepcopy(model).to("cuda:1")
+    ref_model.eval()
+    ref_model.requires_grad_(False)
+
+    reward_model.eval()
+    reward_model.requires_grad_(False)
+
+    layers = model.transformer.h
+    num_layers = len(layers)
+    num_unfrozen = int(0.3 * num_layers)
+    for layer in layers[:-num_unfrozen]:
+        layer.requires_grad_(False)
+
+    adv_fn = GAE(
+        value_network=VmapCritic(critic), gamma=0.99, lmbda=0.95, average_gae=True
+    )
+
+    loss_fn = ClipPPOLoss(actor, critic_head)
+    tdl = get_prompt_dataloader(config, split="train")
+    vdl = get_prompt_dataloader(config, split="valid")
+
+    test_prompt = next(vdl)[:1]
+    estimate_loss = create_loss_estimator(
+        config,
+        reward_model,
+        test_prompt,
+        tokenizer,
+        logger=query_logger,
+        ref_model=ref_model,
+    )
+
+    lr = config["learning_rate"]
+    wd = config["weight_decay"]
+    beta1 = config["beta1"]
+    beta2 = config["beta2"]
+    grad_clip = config["grad_clip"]
+
+    optimizer = torch.optim.AdamW(
+        loss_fn.parameters(), lr=lr, weight_decay=wd, betas=(beta1, beta2)
+    )
+
+    rb = TensorDictReplayBuffer(
+        storage=LazyTensorStorage(config["episode_length"] * config["num_rollouts"]),
+        batch_size=config["ppo_batch_size"],
+        sampler=SamplerWithoutReplacement(),
+    )
+    losses = []
+    rewards = []
+    test_rewards = []
+
+    for i in trange(config["max_iters"]):
+        rb.empty()
+        rollout_rewards = []
+        for _ in range(0, config["num_rollouts"], config["batch_size"]):
+            batch = next(tdl)
+            generated, log_probs, log_ratio = generate(
+                model, batch, ref_model=ref_model
+            )
+            # generate the tensordict structure expected from a rollout using the generated
+            # tokens from the huggingface model
+            td = create_rollout_td(batch, generated, reward_model, log_probs, log_ratio)
+            with torch.no_grad():
+                adv_fn(td)
+            # it's possible we didn't fill the replay buffer in the last iteration if
+            # generation stopped early, so we empty first before repopulating
+            rb.extend(flatten_td(td))
+            rollout_rewards.append(td.get(("next", "reward")).mean().cpu().item())
+        rewards.append(torch.tensor(rollout_rewards).mean().cpu().item())
```

Comment:
why do we call torch.tensor -> mean -> item?
isn't it a tensor already?

Reply:
No, it's a list of floats (see line immediately preceding).

Perhaps we could calculate a rolling average instead or similar to avoid the list appends, but I don't think it's a bottleneck atm.